sdk/src/reserved_account_keys.rs

```diff
@@ -0,0 +1,204 @@
+//! Collection of reserved account keys that cannot be write-locked by transactions.
+//! New reserved account keys may be added as long as they specify a feature
+//! gate that transitions the key into read-only at an epoch boundary.
+
+#![cfg(feature = "full")]
+
+use {
+    crate::{
+        address_lookup_table, bpf_loader, bpf_loader_deprecated, bpf_loader_upgradeable,
+        compute_budget, config, ed25519_program, feature,
+        feature_set::{self, FeatureSet},
+        loader_v4, native_loader,
+        pubkey::Pubkey,
+        secp256k1_program, stake, system_program, sysvar, vote,
+    },
+    lazy_static::lazy_static,
+    std::collections::HashSet,
+};
+
+// Inline zk token program id since it isn't available in the sdk
+mod zk_token_proof_program {
+    solana_sdk::declare_id!("ZkTokenProof1111111111111111111111111111111");
+}
+
+pub struct ReservedAccountKeys;
+impl ReservedAccountKeys {
+    /// Compute the active set of reserved keys based on activated features
+    pub fn active(feature_set: &FeatureSet) -> HashSet<Pubkey> {
```

Comment:
Thinking more about this api, and how it's used in the runtime (in your larger PR), I don't think spitting out the total active list from the `static` vec is the right approach.

The existing api would be functional to handle the existing list of active and pending keys, but doesn't fully satisfy the design approved in [the SIMD](https://github.com/solana-foundation/solana-improvement-documents/blob/main/proposals/0105-dynamic-reserved-accounts-set.md#detailed-design), which says:

> On epoch boundaries, validator implementations should add or remove reserved keys as dictated by feature gated code.

Say someone accidentally deleted an `active` row from `RESERVED_ACCOUNT_KEYS`. That key would be removed from `Bank::reserved_account_keys()` on the next epoch boundary after upgrade, regardless of whether a feature gate had been activated.

Or say we add a `reserved_account_key` to the Bank another way (perhaps it's a key with a dynamic address, like we were discussing on https://github.com/solana-foundation/solana-improvement-documents/pull/106 a while back). On the next epoch boundary, that key would be overwritten and removed from the set. 

For the runtime, I think we need apis to add and remove from the `Bank::reserved_account_keys()` list, rather than replacing the field entirely.

Reply:
The implementation is pretty much the same as how active feature gate keys are determined for each epoch boundary. If someone accidentally deletes an active feature gate, it will no longer be active at the following epoch boundary just like this code. If we add a reserved account key to the Bank in another way they can all be dynamically determined at the epoch boundary and unioned with the static reserved account key set. So I don't really see the point in changing the API unless we expect the set of active reserved keys to change in the middle of an epoch boundary.

Reply:
> The implementation is pretty much the same as how active feature gate keys are determined for each epoch boundary

I don't think that's accurate. `compute_active_feature_set()` begins with active and inactive lists duplicated from the parent Bank.
https://github.com/solana-labs/solana/blob/e0203f22dc83cb792fa97f91dbe6e924cbd08af1/runtime/src/bank.rs#L7276

It is true that if someone accidentally deletes an active feature gate, it will no longer be active at the following epoch boundary ... after reboot. But that is because the `Bank::new_from_fields()` code that runs on restart from snapshot (and starts with a `FeatureSet::default()`, not because `Bank::feature_set` gets wiped and recalculated on the epoch boundary.

I'll concede that we probably have to do something similar for this set on node reboot. But I don't think that's a very strong argument in favor of the existing api.

Reply:
Ah ok I see your point, sorry I misunderstood how epoch boundaries worked for feature gates. I'll make the suggested changes tomorrow.